### PR TITLE
Add types for the faker-schema package

### PR DESCRIPTION
**Pull request by yutro**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/faker-schema/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// Type definitions for faker-schema 1.0
+// Project: https://github.com/knicklabs/faker-schema
+// Definitions by: Yury A Troynov <https://github.com/yutro>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+export function withProbability<T>(value: (() => T) | T, probability?: number): <S>(schema: S) => (T extends () => void ? S : T) | null;
+
+export class Schema<T> {
+    private seed?: number;
+    private blueprint: () => T;
```
Comment (peterblazejewicz) [inline types/faker-schema/index.d.ts:10]: those are private and are overriden via setters only, am I correct? (desribed in readme)

File: types/faker-schema/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    constructor(blueprint: () => T);
```
Comment (peterblazejewicz) [inline types/faker-schema/index.d.ts:11]:
```suggestion
    constructor(blueprint?: () => T);
```
this supports default value, so IMO it's an optional param:
https://github.com/knicklabs/faker-schema/blob/f953b3a6011c288a3743740cef5eb116ff50c12b/src/lib/Schema.js#L6

File: types/faker-schema/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+// Type definitions for faker-schema 1.0
+// Project: https://github.com/knicklabs/faker-schema
+// Definitions by: Yury A Troynov <https://github.com/yutro>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+export function withProbability<T>(value: (() => T) | T, probability?: number): <S>(schema: S) => (T extends () => void ? S : T) | null;
+
+export class Schema<T> {
```
Comment (peterblazejewicz) [inline types/faker-schema/index.d.ts:8]: The T has to extend object/record, as per description: `create one or more pseudo-random records from schema`, where a schema created with `A blueprint (..) function that returns an object literal.`
T cannot be anything:
```ts
const bluePrint = () => (41);
const personSchema = new Schema(bluePrint);
const person = personSchema.makeOne();
person.toFixed();
```
(which is legit TS, no errors)

**Review by sheetalkamat — CHANGES_REQUESTED** [review]
Please address feedback

File: types/faker-schema/index.d.ts (hunk 1 above)
Comment (yutro) [reply types/faker-schema/index.d.ts:10]: So there are no classical setter for `seed `, its being overridden only when makeOne method called with seed parameter.
ex.
https://github.com/knicklabs/faker-schema/blob/f953b3a6011c288a3743740cef5eb116ff50c12b/src/lib/Schema.js#L6
line 25.

File: types/faker-schema/index.d.ts (hunk 3 above)
Comment (yutro) [reply types/faker-schema/index.d.ts]: added now it extends Object :)

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@yutro thanks!

**Comment by yutro** [comment]
@sheetalkamat hey hey :) do you still want me to change something ?

**Comment by yutro** [comment]
@sheetalkamat  hey hey did you have a chance to check my previous comment ?

**Pull request closed without merging**